keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,203 @@
+	}
+	labelPaths := [][]string{
+		{"metadata", "labels"},
+		{"spec", "template", "metadata", "labels"},
+		{"spec", "volumeClaimTemplates[]", "metadata", "labels"},
+		{"spec", "jobTemplate", "metadata", "labels"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "labels"},
+	}
+
+```
+The labels are very static no change is required here. Labels can be used as selectors, and it still will be
+beneficial to have the heritage, operator and instance labels on templated pods.
+
+
+```
+	fieldsToAdd := map[string]string{
+		kudo.PlanAnnotation:            metadata.PlanName,
+		kudo.PhaseAnnotation:           metadata.PhaseName,
+		kudo.StepAnnotation:            metadata.StepName,
+		kudo.OperatorVersionAnnotation: metadata.OperatorVersion,
+		kudo.PlanUIDAnnotation:         string(metadata.PlanUID),
+	}
+	annotationPaths := [][]string{
+		{"metadata", "annotations"},
+		{"spec", "template", "metadata", "annotations"},
+		{"spec", "jobTemplate", "metadata", "annotations"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "annotations"},
+	}
+```
+The annotations are volatile and will/may change with every plan execution. The nested changes to template and jobTemplate
+are historically made because the behavior was modeled after Kustomize.
+
+This KEP proposes to change this: 
+- Attributes will only be applied to the top-level resource, not nested resources ( `spec/template/...`, `spec/jobTemplate/...`)
+- The PlanUIDAnnotation will be completely removed.
+
+The values of the attributes can still be discovered by following the ownerReference of created resources.
+
+### Risks and Mitigations
+
+The proposed change removes the automatic pod restart. It will be a breaking change, operator developers will need
+to adjust the operator to define a group for all used parameters and add their own attribute in the pod spec template
+that triggers the restart. This needs to be documented, and old operators will behave differently with the new KUDO
+version.
+
+## Implementation History
+
+- 2020-03-31 - Initial draft. (@aneumann)
+- 2020-04-14 - Add second Proposal
+- 2020-04-15 - Cleanup, moved first proposal to alternatives section
+
+## Alternatives
+
+### Full dependency graph for used resources and parameters
+
+A better alternative would be to have a full dependency graph of all resources to parameters. This would need to include
+ConfigMaps, Secrets, potentially other pods, etc. With this dependency graph, KUDO could determine when a pod restart
+is necessary and when not, i.e. changing the replica value of a stateful set would not trigger a pod restart, changing
+a value in a ConfigMap that is used by the pod template would trigger a restart.
+An operator would still want to configure if and when pod restarts may be required, as a pod can be aware of changing
+ConfigMaps. 
+
+Calculating the dependency graph would be a very complex undertaking, and may not even be possible in all cases. The
```

Comment:
I'm not sure this is true...  I'm not sure what you may have in mind for a dependency graph... but it is fairly easy to get a list of tasks for a plan and a list of resources for those tasks... and should be easy to get all the params used by those resources (we do this for package verification)... then we could have an attribute on params which indicates it is to be excluded  (inclusion by default) in triggered update.   The automatic hash would be the collection of all params for all resources for that plan that are "included".   I think I like this approach more.   It would be great to reduce the burden on the operator dev.   I'm not a fan (as of yet) with the "grouping" of params

Reply:
Yeah, I've talked with @zen-dog about the complexity of building this dependency graph, and it seems doable, although I feel it might be a bit complex. But I'll expand on this idea a bit more.

Reply:
we should add another approach which is parameter-free 

and we can use the task's rendered resources to build a dependency graph. If the rendered content of CM(resource) is the same we know we don't need to update all dependent resources. But if the content of CM has changed we need to update the dependent resources. 

All resources can have an annotation of the hash of rendered resource on which they depend. And KUDO just updated the hash on each update. 

When I say rendered resources I mean a rendered version of templates, for example, https://github.com/kudobuilder/operators/blob/master/repository/kafka/operator/templates/health-check.yaml where we can control the content. 
Because with kubernetes objects any small status update will change the hash.

Reply:
That sounds workable. I have some thoughts about resources about resources that are deployed by different plans, but that should be solvable. I'll flesh that out and add it to the KEP

Reply:
@zmalik I've updated the KEP with your proposal, and I think that looks like the best solution so far.